Pass in the appropiate sequence no. for "wait" op

shengfukevin opened the pull request:
Summary:
In a ProcessGroupNCCL, there are two different sequence no. for collective and p2p op. "wait" op need to know what type of op it waits.

In this DIFF, I changed RECORD_PARAM_COM interface to add a new entry "isP2P", this is a boolean varaible to flag if the op is a point to point op or not.

Differential Revision: D61882985


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

GSSBMW commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2448,7 +2450,7 @@ c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
       rank,
       opType,
-      seqCollective_,
+      isP2POp(opType) ? seqP2P_++ : seqCollective_++,
```
Why `++` here?

GSSBMW commented on torch/csrc/distributed/c10d/ParamCommsUtils.hpp:
```diff
@@ -96,6 +96,7 @@ class TORCH_API ParamCommsDebugInfo : public c10::DebugInfoBase {
 
 #define RECORD_PARAM_COMMS(                                                    \
     seq,                                                                       \
+    isP2P,                                                                     \
```
How about make seq be a tuple of (seq_id, is_p2p), but not a stand alone integer? Then make `RECORD_PARAM_COMM` and `RECORD_PARAM_COMM_DATA` have the same trace fromat, exept the first tensor. Otherwise, replayer needs to branch positional parsing.
After that, we will use `(seq_id, is_p2p, pg_id)` as key to match comm op and wait.

GSSBMW commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4009,6 +4011,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::barrier(const BarrierOptions& opts) {
   RECORD_PARAM_COMMS(
       static_cast<int>(
           this->getSequenceNumberForGroup() + 1), // seq + 1 to match collective
```
seems we should remove the function `getSequenceNumberForGroup()`, but use `seqCollective_` directly.
1. func name can't represent it returns collective sequence id;
2. keep conssistant with P2P, which use `seqP2P` directly;

Or we can reimplement `getSequenceNumberForGroup()` to:
```
uint64_t ProcessGroupNCCL::getSequenceNumberForGroup(OpType typ, bool batchP2P = false) {
  return isP2P(typ, batchP2P) ? seqP2P_ : seqCollective_;
}
```

shengfukevin replied:
getSequenceNumberForGroup is a virtual function defined in root class BackEnd. P2P seems only used in ProcessGroupNCCL. Add a new input parameter probably does not make sense to other sub-classes.

GSSBMW replied:
got it. Then how about using `seqCollective_` directly?

GSSBMW commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2428,7 +2430,7 @@ c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
       rank,
       opType,
-      seqCollective_,
+      isP2POp(opType) ? seqP2P_ : seqCollective_,
```
according to [isP2POp](https://github.com/pytorch/pytorch/blob/1011e0ae9804baca9ef06abfeb6903f15459d346/torch/csrc/distributed/c10d/ProcessGroup.cpp#L65), batched P2P is not P2P op. So we need to check coalesing here.

shengfukevin replied:
batched P2P is implemented through Coalescing. The id we want to pick is based on if it is either a single P2P op/coalesced P2P op or a single collective op/coalesced collective op, does not matter it is coalesced or not.

GSSBMW replied:
Ok. If it's coalesed p2p, `seqCollective_` is bumped up in `startCoalesing()` and we pick `seqCollective_` here.
Not sure whether this will be changed/rectified according to this [comment](https://github.com/pytorch/pytorch/issues/134833#issuecomment-2322735566) from @fduwjj

fduwjj reviewed (approved):
LGTM

GSSBMW commented:
Let's wait for a conclusion of this [issue](https://github.com/pytorch/pytorch/issues/134833)?

shengfukevin committed:
Pass in a flag for "wait" op to indicate if it is a point to point op (#134619)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134619

In a ProcessGroupNCCL, there are two different sequence no. for collective and p2p op. "wait" op need to know what type of op it waits.

In this DIFF, I changed RECORD_PARAM_COM interface to add a new entry "isP2P", this is a boolean varaible to flag if the op is a point to point op or not.

Test Plan: buck2 run mode/opt kineto/libkineto/fb/integration_tests:pytorch_execution_trace_integration_test

Reviewed By: c-p-i-o

Differential Revision: D61882985

GSSBMW commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4263,7 +4264,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::send(
   RECORD_PARAM_COMMS_DATA(
       static_cast<int>(
-          this->getSequenceNumberForGroup() + 1), // seq + 1 to match collective
+          this->seqP2P_ + 1), // seqP2P_ + 1 to match pointToPoint op
```
seems better to use tuple `(seq, is_p2p)` for `RECORD_PARAM_COMMS_DATA`. Otherewise, replayer needs to copy the logic of PyTorch to check `collectiveName_`. Keep consistant pattern would be better.

GSSBMW commented:
@shengfukevin Please refer to this PR (https://github.com/pytorch/pytorch/pull/134578) or use it directly. Thanks!
